irobot_events_executor/src/rclcpp/executors/events_executor/events_executor.cpp

```diff
@@ -64,7 +64,6 @@ EventsExecutor::spin()
   RCPPUTILS_SCOPE_EXIT(this->spinning.store(false); );
 
   timers_manager_->start();
-  RCPPUTILS_SCOPE_EXIT(timers_manager_->stop(); );
```

Comment:
why removing this?

Reply:
I explained in the commit, 
when `RCPPUTILS_SCOPE_EXIT` happens, the `timers_manager_` has already been destroyed.
So calling `timers_manager_->stop()` will try to lock an already destroyed mutex, so it asserts(0).
The destructor of the `timers_manager_` already calls `stop()` so I'm not sure we need this line I'm removing here.

Reply:
This the example to reproduce:
```
  auto exec = std::make_unique<rclcpp::executors::EventsExecutor>();
  exec->add_callback_group(m_callback_group, node_base->get_node_base_interface());
  std::thread t = std::thread([&exec](){exec->spin();});
  exec.reset();
```
The `exec.reset();` destroys the executor, thus also the TimersManager.
When `exec->spin()` exits, it has already been destroyed:
```
exec.reset()
~EventsExecutor
~EventsExecutorEntitiesCollector
~TimersManager()
EventsExecutor::spin() - EXIT
RCPPUTILS_SCOPE_EXIT
~Executor: ~interrupt_guard_condition_
~Executor: ~shutdown_guard_condition_
rclcpp::shutdown();
rclcpp::shutdown() - DONE

```

Reply:
I think just checking if exists before would be enough.

Reply:
Shouldn't the destructor first interrupt the thread and then destroy everything?

Reply:
So, 
```
Thread1 -> spin()
Thread2-> destroy stuff
```

Reply:
But we could modify things to do as you say:
1. On destructor, first interrupt the thread and wait until is finished
2. Then keep destroying things

Reply:
I mean that we could modify the destructor of the `EventsExecutor` such that it blocks until `spin()` terminates, but this likely complicates things a lot.

It's likely easier to just do the check, although I'm not sure if we are fully safe with this change alone.

Reply:
The check is not enough unless we add a mutex:
```
  RCPPUTILS_SCOPE_EXIT( if(timers_manager_) { timers_manager_->stop(); });
```
Could happen that the `timers_manager_` still exists during the `if`, but then is destroyed before
```timers_manager_->stop();```

In my tests, if I manually reset the timers like
```
EventsExecutor::~EventsExecutor()
{
  timers_manager_.reset();
  spinning.store(false);
}
``` 
the 
```
  RCPPUTILS_SCOPE_EXIT( if(timers_manager_) { timers_manager_->stop(); });
```
works fine without need for any mutex

Reply:
Here I wait for `spin` to exit before destroying stuff: 05cd1a8aeba7b1ed7d6307ca352463e0a28dbf69

Reply:
A couple of comments:
 - `spin_has_finished_` should be set to false using `RCPPUTILS_SCOPE_EXIT` otherwise it won't work when the executor throws an exception
 - do we need this variable? could we rather check whether the timers-manager has been stopped?
 - do we really need polling? 100ms seems a huge amount of time (it means that in a system with 50 executors we could wait 5 seconds just to shut those down. can we use a condition variable?
 - whats' the purpose of triggering the shutdown guard condition?

Reply:
- How can you check that the timers-manager has been stopped?
- Trigger the guard condition is to exit the wait on the queue, otherwise it hangs forever:
https://github.com/irobot-ros/events-executor/blob/8c47a89b63ab5a502e9a1ed8bbcb3ad3c9cfab7c/irobot_events_executor/include/rclcpp/executors/events_executor/simple_events_queue.hpp#L69
The trigger pushes an event into the queue. And since we set `spinning` to false, `EventsExecutor::spin()` will exit the loop
- I can set `spin_has_finished_` in `RCPPUTILS_SCOPE_EXIT`

Reply:
We could add a method `is_stopped` in the timers-manager

Reply:
Done in 2d670f78a1e62ce8c26483ce6720d393364c512a

Reply:
Corrected in 44ef4beaa13649e04438b3582d4d420420f7d4d7